### PR TITLE
[VEN-2984] Flash loan implementation

### PR DESCRIPTION
**Pull request by Debugger022**
## Description

This PR introduces the implementation of flashloan functionality for isolated pools. Flashloan enable users to borrow assets without collateral, provided the loan is repaid within the same transaction. This enhancement includes the following changes:

Flashloan Implementation: Added flashloan capability to isolated pools, allowing borrowing and repayment in a single transaction.

Mock Receiver Contract: Introduced a mock receiver contract for testing purposes, which handles the flashloan receipt and repayment.

Test Cases: Comprehensive unit tests for the flashloan logic, including tests for successful execution, edge cases, and failure scenarios.

Deployment Script: Updated the deployment script to accommodate the new flashloan functionality for isolated pools.

Contract Updates: Modified relevant contracts to integrate the flashloan feature while ensuring compatibility with existing logic and maintaining security checks.

### CODE REVIEW

File: contracts/Comptroller.sol [hunk 1]
```diff
@@ -1099,7 +1219,11 @@ contract Comptroller is
             revert MarketAlreadyListed(address(vToken));
         }
 
-        require(vToken.isVToken(), "Comptroller: Invalid vToken"); // Sanity check to make sure its really a VToken
+        if (!vToken.isVToken()) {
+            revert InvalidVToken();
+        }
+
+        // require(vToken.isVToken(), "Comptroller: Invalid vToken"); // Sanity check to make sure its really a VToken
```
Comment (web3rover) [inline contracts/Comptroller.sol:1226]: We can remove this commented line

File: contracts/Comptroller.sol [hunk 2]
```diff
@@ -1133,7 +1257,10 @@ contract Comptroller is
         uint256 numMarkets = vTokens.length;
         uint256 numBorrowCaps = newBorrowCaps.length;
 
-        require(numMarkets != 0 && numMarkets == numBorrowCaps, "invalid input");
+        if (numMarkets == 0 || numMarkets != numBorrowCaps) {
+            revert InvalidInput();
+        }
+        // require(numMarkets != 0 && numMarkets == numBorrowCaps, "invalid input");
```
Comment (web3rover) [inline contracts/Comptroller.sol:1263]: We can remove this commented line

File: contracts/Comptroller.sol [hunk 3]
```diff
@@ -1554,7 +1687,10 @@ contract Comptroller is
      * @param paused The new paused state (true=paused, false=unpaused)
      */
     function _setActionPaused(address market, Action action, bool paused) internal {
-        require(markets[market].isListed, "cannot pause a market that is not listed");
+        if (!markets[market].isListed) {
+            revert MarketNotExist();
+        }
+        // require(markets[market].isListed, "cannot pause a market that is not listed");
```
Comment (web3rover) [inline contracts/Comptroller.sol:1693]: We can remove this commented line

File: contracts/Comptroller.sol [hunk 4]
```diff
@@ -919,6 +955,81 @@ contract Comptroller is
...
+     * @custom:reverts
+     *      - Reverts with `InvalidFlashLoanParams()` if parameter checks fail.
+     *      - Reverts with `FlashLoanNotEnabled(asset)` if flashLoans are disabled for any of the requested assets.
+     *      - Reverts with `ExecuteFlashLoanFailed` if the receiver contract fails to execute the operation.
+     *      - Reverts with `InsufficientReypaymentBalance(asset)` if the repayment (amount + fee) is insufficient after the operation.
+     */
+    function executeFlashLoan(
+        address receiver,
+        VTokenInterface[] calldata assets,
+        uint256[] calldata amounts,
```
Comment (web3rover) [inline contracts/Comptroller.sol:983]: It would be more appropriate to rename `assets` to `vTokens` and `amounts` to `underlyingAmounts`

Comment (Debugger022) [reply contracts/Comptroller.sol]: https://github.com/VenusProtocol/isolated-pools/pull/460/commits/79f792d23be54291c106d74909366b0d87f5fd22

**Comment by fred-venus** [comment]
close for now, following current design flashloan should only be available in core pool
